src/Lykke.blue.Service.ReferralLinks/Controllers/ReferralLinksController.cs

```diff
@@ -0,0 +1,480 @@
+
+                var validationError = ValidateClaimRefLinkAndRequest(refLink, request, refLinkId);
+                if (!String.IsNullOrEmpty(validationError))
+                {
+                    return await LogAndReturnBadRequest(request, ControllerContext, validationError);
+                }
+
+                var claims = (await _referralLinkClaimsService.GetRefLinkClaims(refLink.Id)).ToList();
+
+                var alreadyClaimedByThisRecipient = claims.Any(c => c.RecipientClientId == request.RecipientClientId);
+                if (alreadyClaimedByThisRecipient)
+                {
+                    return await LogAndReturnBadRequest(request, ControllerContext, $"Link already claimed by client id {request.RecipientClientId}");
+                }
+
+                bool shouldReceiveReward = await ShoulReceiveReward(claims, refLink);
+
+                var newRefLinkClaimRecipient = await CreateNewRefLinkClaim(refLink, request.RecipientClientId, shouldReceiveReward, true);
+
+                if (shouldReceiveReward)
+                {
+                    var newRefLinkClaimSender = await CreateNewRefLinkClaim(refLink, refLink.SenderClientId, true, false);
+
+                    var transactionRewardSender = await _exchangeService.TransferRewardCoins(refLink, false, refLink.SenderClientId, ControllerContext.GetControllerAndAction());
+                    await SetRefLinkClaimTransactionId(transactionRewardSender, newRefLinkClaimSender);
+
+                    var transactionRewardRecipient = await _exchangeService.TransferRewardCoins(refLink, request.IsNewClient, request.RecipientClientId, ControllerContext.GetControllerAndAction());
+                    await SetRefLinkClaimTransactionId(transactionRewardRecipient, newRefLinkClaimRecipient);
+
+                    if (transactionRewardSender.IsOk() && transactionRewardRecipient.IsOk())
+                    {
+                        return Ok
+                            (
+                                new ClaimRefLinkResponse
+                                {
+                                    TransactionRewardSender = transactionRewardSender.TransactionId,
+                                    TransactionRewardRecipient = transactionRewardRecipient.TransactionId
+                                }
+                            );
+                    }
+                    return await LogAndReturnInternalServerError(request, ControllerContext, $"TransactionRewardRecipientError: Code: {transactionRewardRecipient.Code}, {transactionRewardRecipient.Message}; TransactionRewardSenderError: Code: {transactionRewardSender.Code}, {transactionRewardSender.Message}");
+                }
+
+                return Ok(new ClaimRefLinkResponse());
+            }
+            catch (Exception ex)
+            {
+                await LogError(request, ControllerContext, ex);
+                return NotFound(ErrorResponseModel.Create($"{ex.Message}.{ex.InnerException?.Message}."));
```

Comment:
I think this is InternalServerError. And we should hide error details from client.

Reply:
Client app does not show these error messages to the client.

Reply:
Removed error details